pkg/migration/spec.go

```diff
@@ -0,0 +1,17 @@
+// Package migration provides implementation for CR migration from Giant Swarm
+// flavor to upstream compatible CAPI.
+package migration
+
+type Migrator interface {
+	// IsMigrated performs check to see if given cluster has been already
+	// migrated.
+	IsMigrated() (bool, error)
+
+	// Prepare executes preparatory migration actions such as transforming
+	// existing CRs into upstream compatible format and creating missing CRs.
+	Prepare() error
+
+	// Migrate performs final execution which shifts reconciliation to upstream
+	// controllers.
+	Migrate() error
+}
```

Comment:
Do you think you can provide some generic execution logic for this interface as well?

Reply:
Since I expect this operator to be run with single instance per provider (i.e. the same instance won't perform migration for aws and azure), the correct implementation should come from initial wiring.

Then my thinking was roughly something as follows:

```go

alreadyMigrated, err := migrator.IsMigrated()
if err != nil {
    return microerror.Mask(err)
}

if alreadyMigrated {
    // Migration already performed. No need to do anything.
    return nil
}

err = migrator.Prepare()
if err != nil {
   return microerror.Mask(err)
}

err = migrator.Migrate()
if err != nil {
   return microerror.Mask(err)
}
```

The `Migrator` implementation should probably be written in the way that it ensures that ordering of execution. Final migration shouldn't be triggered before preparation. I originally had only one function `Migrate()` but then started thinking that if cloud infrastructure requires changes, there might be a need for separated intermediate steps. Those could be implemented inside the single function call as well though. Not sure which one is better approach?

Reply:
Hmm... Gotta give this few more minutes of thought. The cluster identifier needs connection in-between, but it's not good to be in parameter. Maybe small factory to instantiate the migrator per cluster.

Reply:
Maybe we also need a `migrator.Cleanup()` step to remove leftovers?

Reply:
Edited the snippet above.

Reply:
> Maybe we also need a migrator.Cleanup() step to remove leftovers?

Hmm.. That makes sense, though I need to refactor the interface a bit more.

Reply:
Added two more functions to interface + refined above snippet.